### PR TITLE
Use once_cell to replace OnceSlot

### PR DESCRIPTION
**Pull request by cuviper**
There was a comment "TODO: This should be its own crate", but a similar
type already exists in the `once_cell` crate.

### CODE REVIEW

**Comment by Lucretiel** [comment]
I haven't set up CI yet; can you confirm that tests & clippy both pass?

**Comment by cuviper** [comment]
Yes, test and clippy are clean here.

**Comment by cuviper** [comment]
FWIW, I would also suggest trimming the `crossbeam` dependency to just `crossbeam-channel`, but there's no _functional_ change in that. It's up to you.

**Comment by Lucretiel** [comment]
Great, thank you! I briefly looked at `once_cell`, but their front page implied that the only interface was `set` and `get() -> Result<&T, ()>`, and I didn't want that overhead.

Yeah, I'm happy to trim that crossbeam dependency, thank you.

**Comment by cuviper** [comment]
> I briefly looked at `once_cell`, but their front page implied that the only interface was `set` and `get() -> Result<&T, ()>`, and I didn't want that overhead.

cc @matklad on that bad first impression.

File: src/once_slot.rs [hunk 1]
```diff
@@ -1,58 +0,0 @@
...
-        // The read is safe because MaybeUninit doesn't destruct its contents.
-        if self.once.is_completed() {
-            drop(unsafe { (*self.slot.get()).as_ptr().read() });
-        }
-    }
-}
-
-// Safety: our interface ensures that synchronized access to the inner T is
-// well-controlled
-unsafe impl<T: Sync> Sync for OnceSlot<T> {}
```
Comment (matklad) [inline src/once_slot.rs:58]: Note that this is unsound, `T: Send` is also required. `T` can be created by one thread that calls `get_or_init` and dropped by another thread, which owns the `OnceSlot`.

Comment (Lucretiel) [reply src/once_slot.rs:58]: I'm not sure I follow; the value's ownership never changes after it's created, does it?

Comment (matklad) [reply src/once_slot.rs:58]: Yeah, let me clarify a bit

```
struct SyncNotSend { ... };
impl Sync for SyncNotSend {}
impl !Send for SyncNotSend {}

let slot: OnceSlot<SyncNotSend> = OnceSlot::new();
crossbeam::scope(|scope| {
   scope.spawn(|_| {
       let obj = SyncNotSend::new(); // we create this type on the nested thread
       slot.get(|| obj); // move non-send type into slot
   })
});

drop(slot); // here, we drop the `SyncNotSend` value on another thread
```

Here, we create value on the one thread, and drop it on the other. It is possible to create a real "channel" for Sync!Send types this way if, for example, you stuff the value into some thread-local on Drop.

Comment (Lucretiel) [reply src/once_slot.rs:58]: Ahhhh I see, that does make sense. Thanks for the heads-up!

**Comment by Lucretiel** [comment]
> > I briefly looked at `once_cell`, but their front page implied that the only interface was `set` and `get() -> Result<&T, ()>`, and I didn't want that overhead.
> 
> cc @matklad on that bad first impression.

I like that the simplest calls are front and center, but I'm also used to the FnOnce style interface used by Rust's Once (and also Go's Once) so that's the first thing I look for

**Pull request merged**